hazelcast/src/main/java/com/hazelcast/sql/impl/expression/math/AbsFunction.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.math;
+
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.expression.UniExpressionWithType;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.converter.Converter;
+
+import static com.hazelcast.sql.impl.expression.math.ExpressionMath.DECIMAL_MATH_CONTEXT;
+
+public class AbsFunction<T> extends UniExpressionWithType<T> implements IdentifiedDataSerializable {
+    @SuppressWarnings("unused")
+    public AbsFunction() {
+        // No-op.
+    }
+
+    private AbsFunction(Expression<?> operand, QueryDataType resultType) {
+        super(operand, resultType);
+    }
+
+    public static AbsFunction<?> create(Expression<?> operand, QueryDataType resultType) {
+        return new AbsFunction<>(operand, resultType);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T eval(Row row, ExpressionEvalContext context) {
+        Object operandValue = operand.eval(row, context);
+
+        if (operandValue == null) {
+            return null;
+        }
+
+        return (T) abs(operandValue, operand.getType(), resultType);
+    }
+
+    private static Object abs(Object operand, QueryDataType operandType, QueryDataType resultType) {
+        Converter operandConverter = operandType.getConverter();
+
+        switch (resultType.getTypeFamily()) {
+            case TINYINT:
+                return (byte) Math.abs(operandConverter.asTinyint(operand));
```

Comment:
Are conversions necessary here, if all the types are already should be inferred statically by Calcite?

Reply:
IMO there is no much difference. Converters are safer than relying on Calcite. In most cases, the return type will be the same, so the converter will not do anything except for cast from Object to a concrete type.

Reply:
The whole idea of switching to static typing was to assign valid types statically and avoid runtime conversions as much as possible. Defensive usage of converters at runtime doesn't make it safer, it makes it more uncertain. If there are some bugs with type inference and/or implicit casts behavior for parameters of math functions, usage of converters will hide them, since converters are readily converting everything. And there are no tests similar to `UnaryMinusTest` verifying the actual types assigned to parameters of math functions to give confidence.

Reply:
There were no idea to avoid runtime conversions. This specific expression works in a way that it can tolerate different inputs for so long, as values are convertible. It is not concerned with Calcite integration.